guides/source/sign_up.md

```diff
@@ -0,0 +1,1652 @@
+
+```ruby#14-17
+module Authentication
+  extend ActiveSupport::Concern
+
+  included do
+    before_action :require_authentication
+    helper_method :authenticated?
+  end
+
+  class_methods do
+    def allow_unauthenticated_access(**options)
+      skip_before_action :require_authentication, **options
+    end
+
+    def unauthenticated_access_only(**options)
+      allow_unauthenticated_access **options
+      before_action -> { redirect_to root_path if authenticated? }, **options
+    end
+
+    # ...
+```
+
+The `unauthenticated_only_access` class method can be used in any controller where we want to restrict actions to unauthenticated users only.
+
+We can then use this method at the top of `SignUpsController`.
+
+```ruby#2
+class SignUpsController < ApplicationController
+  unauthenticated_access_only
+
+  # ...
+end
+```
+
+### Rate Limiting Sign Up
+
+Our application will be accessible on the Internet so we're bound to have malicious bots and users trying to spam our application. We can add rate limiting to sign up to slow down anyone submitting too many requests.
+
+Rails makes this easy with the [`rate_limit`](https://api.rubyonrails.org/classes/ActionController/RateLimiting/ClassMethods.html) method in controllers.
+
+```ruby#3
+class SignUpsController < ApplicationController
+  unauthenticated_only_access
+  rate_limit to: 10, within: 3.minutes, only: :create, with: -> { redirect_to sign_up_path, alert: "Try again later." }
+
+  # ...
+end
+```
+
+This will block any form submissions that happen more than 10 times within 3 minutes.
+
+Editing Passwords
+-----------------
+
+Now that users can login, we should create a few places for users to update their profile, password, email address and other settings.
+
+### Using Namespaces
+
+The Rails authentication generator already created `app/controllers/passwords_controller.rb` and uses this controller for password resets. This means we need to use a different controller for editing passwords of authenticated users.
+
+To prevent conflicts, we can use a feature called namespaces. A namespace organizes routes, controllers, and views into folders and helps prevent conflicts like our two passwords controllers.
+
+We'll create a namespace called "Settings" to separate out the user and store settings from the rest of our application.
+
+In `config/routes.rb` we can add the Settings namespace along with a resource for editing passwords:
+
+```ruby
+namespace :settings do
+  resource :password, only: [:show, :update]
+end
+```
+
+This will generate a route for `/settings/password` for editing the current user's password and is separate from the password resets routes at `/password`.
+
+### Adding the Namespaced Passwords Controller & View
+
+Namespaces also move controllers into a matching module in Ruby. This controller will be in a `settings` folder to match the namespace.
+
+Let's create the controller at `app/controllers/settings/passwords_controller.rb` and start with the `show` action.
```

Comment:
Does it need instructions showing how to create this controller? Open to discuss since at this point we might assume users know. (This also relates to the 2 times I mentioned it above,)

Reply:
I think it's safe to assume they went through the Getting Started guide and understand this?